### PR TITLE
Fixed a copy-reference bug with CompleteEntity tag updates

### PR DESCRIPTION
**Pull request by lucaspcram**
### Description:

The `CompleteEntity` methods `withAddedTag`, `withReplacedTag`, and `withRemovedTag` were broken when multiple `ChangeAtlas`es were chained together. This was due to a pointer copy in the `addNewTag` and `removeTag` methods, which caused disparate `CompleteEntities` to share the same underlying tag map. We now make copies of the maps, so updating one entity will not inadvertently update an unrelated `CompleteEntity` that happened to initialize itself with the same underlying map.

### Potential Impact:

May impact code that relied on the previous buggy behavior.

### Unit Test Approach:

Added `testStackedChangeAtlasesWithTagChanges` to `ChangeAtlasTest` which demonstrates the buggy behavior.

### Test Results:

All tests now pass.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

**Comment by yazad3** [comment]
Great job @lucaspcram! Thank you for the fix.

**Comment by lucaspcram** [comment]
> Please add a test case that does the same set of change but without the context. Ideally with and without `context` being passed to the `CompleteEntity`.

@yazad3 Check the PR I added a test for this. It looks to be OK.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java [hunk 1]
```diff
@@ -189,6 +190,19 @@ public boolean isShallow()
...
+    @Override
+    public void setTags(final Map<String, String> tags)
+    {
+        if (tags != null)
+        {
+            this.tags = new HashMap<>(tags);
+        }
+        else
+        {
+            this.tags = new HashMap<>();
```
Comment (lucaspcram) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java]: @matthieun In the case that the parameter `tags` is `null`, should we be setting `this.tags` to an empty map? Or should we be doing `this.tags = null;`. Curious to hear your thoughts.

**Comment by yazad3** [comment]
@lucaspcram - approved the changes. Thank you.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java (hunk 1 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java]: I believe we should keep it null in that case, as shallow version contain null members.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java [hunk 2]
```diff
@@ -219,7 +220,7 @@ public void setExplicitlyExcludedOutEdgeIdentifiers(final Set<Long> edges)
     @Override
     public void setTags(final Map<String, String> tags)
     {
-        this.tags = tags;
+        this.tags = tags != null ? new HashMap<>(tags) : null;
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteNode.java]: 👍

**Pull request merged**